rfd/0226-self-healing.md

```diff
@@ -0,0 +1,382 @@
+conveniently allows you to register custom policies[^2] by implementing the
+balancer interface[^4]. This policy will enable gRPC health checking by default
+and use the streaming API to ensure health checks can scale to a large number of
+clients.
+
+To avoid introducing too much new behavior we will base our implementation off of
+the existing `pick_first` policy[^5]. The primary difference will be when the existing
+connection enters the `TRANSIENT_FAILURE` state[^6] new connection attempts will be
+made until either the existing connection returns to `READY` or the new connection
+reaches the `READY` state.
+
+All connection attempts should use a backoff policy with jitter to avoid a thundering
+heard problem.
+
+If the new connection reaches the `READY` state it will replace the old connection
+and the old connection will be shutdown.
+
+#### Health Failure vs Network Failure
+
+We will differentiate between a connection failing due to the service being unhealthy
+and a connection failing due to a network issue.
+
+A connection failing due to a network issue can simply create a new connection.
+
+A connection failing due to an unhealthy service should continue to send RPCs to the
+unhealthy service until a new connection to a healthy service can be established.
+
+This allows for a partially degraded auth service to attempt to service RPCs. This
+preservers todays behavior where all requests get sent to the auth regardless of its
+health.
+
+Once a new connection is established the old connection will be drained gracefully
+and new RPCs will be sent over the new connection.
+
+Any open RPCs on the old connection will be allowed to run to completion.
+
+#### Policy Configuration Discovery
+
+We will explore implementing a gRPC endpoint which will allow for discovering
+new load balancing policies via the connected auth server. This endpoint will be
+called immediately after connecting and the initial health checking is complete.
+
+As a backup plan clients could also discover the new load balancing policy via the
+`/webapi/ping` endpoint. However the gRPC endpoint is the preferred solution.
+
+It will be set in the JSON response as follows:
+
+```json
+{
+    "grpc_client_lb_policy": {
+        "loadBalancingConfig": [{"teleport_pick_healthy": {}}],
+        "healthCheckConfig": {
+            "serviceName": ""
+        }
+    }
+}
+```
+
+This configuration supports adding additional load balancer policies in the future.
+Agents should prefer to use the first known policy listed in the loadBalancingConfig
+array. Currently the `teleport_pick_healthy` configuration is an empty json object but
+fields could be added in the future.
+
+This can be configured on an auth server by setting the environment variable
+`TELEPORT_UNSTABLE_GRPC_CLIENT_LB_POLICY`. When not specified clients will continue
+using the default `pick_first` policy. This will allow us to opt-in to this new
+behavior.
+
+### Proxy Reconnects
+
+#### Background on reversetunnels
```

Comment:
@okraport how do the changes that you are currently working on to make reverse tunnel clients aware of networking issues effect this proposal?